cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -80,6 +80,134 @@ const FunctionDoc list_value_length_doc{
      "Null values emit a null in the output."),
     {"lists"}};
 
+template <typename Type, typename IndexType>
+Status ListElementArray(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ListArrayType = typename TypeTraits<Type>::ArrayType;
```

Comment:
`ListElementArray` is a function representing the `Exec` method but `ListElementScalar` is a struct with an actual `Exec` method. I suggest to use the same approach as other kernels and use a class with an `Exec` method, where `Exec` dispatches the corresponding processing based on input types. With this structure this kernel can be extended to support different combinations of inputs shapes (e.g., scalar-scalar, scalar-array, etc).

Given that this is a scalar kernel

Reply:
We're registering two separate kernels so there's no need to do dispatch here. We could make the struct a function instead, though, for consistency.

Reply:
There are many kernels using only a function for the Exec logic, in this I needed to wrap the `ListElementScalar` just because the `GenerateInteger` required so. I would do that wrapper for `ListElementArray` only when we need it and not before.

Reply:
Maybe a future related-kernel to add is to get a `List<Array>` given multiple element indices, but this out-of-scope.